lua/lean/infoview.lua

```diff
@@ -1,6 +1,4 @@
-local lean3 = require('lean.lean3')
-
-local M = {_infoview = nil, _opts = {}}
+local M = {_infoviews = {}, _infoviews_open = {}, _opts = {}}
```

Comment:
Sorry for still working through things here, but one question -- can you elaborate on your thinking for having separate `_infoviews_open` here, which I think relates to functionality that persists infoview windows after they're semi-closed, right?

What benefit is there in supporting that extra state / how do you see it being used?

Reply:
Good question - it may be wise to refactor this to instead let `_infoviews` have two fields, `data: { win: ..., buf: ...}` and `open`, and remove `_infoviews_open` - let me know if you'd like this in this PR.

If we didn't have a state indicating closure, I'm not sure how we would distinguish between an infoview simply not having been created yet (in which case `_infoviews = nil`, and we _would_ want to open it) and having been explicitly closed.

Reply:
Actually, after the refactor above, we may be able to indicate this by setting `_infoviews[key].data = nil` when closed and `_infoviews[key] = nil` when erased - though doing this might be a little bit less readable.

Reply:
OK I see where you might have been confused here - I moved the infoview opening code to the `M.update()` method because I thought it was necessary for this, but looking back it should be fine to move it back to `M.open()`, and add a config setting `infoview_open_default` to control whether new source windows automatically open it. In this case, I should be able to use whether or not `_infoviews[key] = nil` as an indicator of whether it is open. Let me know if you'd like this in this PR.

Reply:
Right something like that is I think exactly what's confusing me yeah. I think probably it's reasonable to do on top of what's here, yeah, since I'm still trying to wrap my head around things, so I suspect doing it that way will actually simplify things a bit and make them more similar to the original diff, so worth doing I suspect. Let me first just push a simple style commit but then think go for it!

Reply:
Awesome, yeah I think I've finished and yeah it does look a lot cleaner :P

Reply:
Ok actually here are my updated thoughts on this - it may be better just to do the initial refactor I mentioned above, because we're eventually going to have to add persistent state to infoviews between open/close anyways when we implement e.g. pinnable/pausable messages. At that point we would be forced to have a new state field indicating open/closed. Thoughts?

Reply:
I don't think I follow what you mean when you're talking about persistent state yet -- I still don't understand why we don't always just want to completely get rid of an infoview when it's closed -- for pinnable messages it seems to me all you do is temporarily/permanently unhook calling `update` on a window, and then it will stay. Maybe I'm missing something still though.

But I want to get these set of incremental changes merged from this PR before doing more functionality, unless the implementation simplifies.

Reply:
This makes it more similar to VSCode, where you can close the webview without having your set of pinned messages deleted.

This will be one of the aspects of this that seems "necessary" to make this a practical tool. For example, say you're editing Lean and LaTeX side-by side. The infoview is useful for getting an idea of how to interpret the Lean code, but once you understand it you'll probably want to get rid of the infoview for more screen space as you work on your LaTeX code. Now if there's a detail that you missed, you'll want to open the infoview again to see your pinned messages without them having been deleted.

Again, the code we have now should be good for a merge, but I'm pretty certain we'll want to implement persistence in the future.

Reply:
OK, let's take things step by step, if/when we get there we can make the implementation more complex, but for now I think we should stick to the simplest thing that solves the current desired functionality, which is:

* don't error when closing the window manually (#20)
* allow one-per-tab or one-per-window infoviews
* properly look for leanpkg.toml relative to the file that's open (#13)
* support lean 3 and lean 4 buffers simultaneously (#33)
* properly enforce `max_width` even when windows are closed/resized

all of which are in this PR (and which I'm trying to merge one by one essentially).

If after that we need a more complex implementation we can evolve into it.